saml-authentication-server/src/main/java/jetbrains/buildServer/auth/saml/plugin/SamlAuthenticationScheme.java

```diff
@@ -196,6 +210,68 @@ private boolean matchPostfixes(String username, String allowedPostfixes) {
         return false;
     }
 
+    @NotNull
+    private void processGroups(@NotNull SUser user, @NotNull String groups, Boolean removeUnassignedGroups) {
+
+        // Get a list of TeamCity groups
+        Collection<SUserGroup> teamcityGroups = userGroupManager.getUserGroups();
+
+        // Get a lower-cased list of users current groups
+        List<String> usersCurrentGroups = user.getUserGroups().stream()
+                .filter(g -> !g.getKey().equals("ALL_USERS_GROUP")) // We don't want to remove the 'ALL_USERS_GROUP'
+                .map(g -> g.getKey().toLowerCase())
+                .collect(Collectors.toList());
+        LOG.debug(String.format("Users current groups = '%s'", usersCurrentGroups));
+
+        // Split the 'groups' string, lowercase and trim empty results
+        List<String> usersAssignedGroups = Arrays.stream(groups.split(", "))
+                .map(String::trim)
+                .map(String::toLowerCase)
+                .filter(StringUtil::isNotEmpty)
+                .collect(Collectors.toList());
+        LOG.debug(String.format("Users assigned groups from SAML response: '%s'", usersAssignedGroups));
+
+        // What groups to add and what groups to remove
+        List<String> groupsToAdd = new ArrayList<>(CollectionUtils.subtract(usersAssignedGroups, usersCurrentGroups));
+
+        // Add any new groups
+        groupsToAdd.forEach(addGroup -> {
+           SUserGroup tcGroup = teamcityGroups.stream()
```

Comment:
Not the best way to lookup things - you literally iterating the collection over and over again (and do it twice - here and below). Instead just make teamcityGroups a map (Collectors.toMap(...) with lowercased group key) and use .containsKey

Reply:
Heh, this is my first time writing Java, so thanks for the pointers.

I'll rework.

Reply:
I think I've addressed in [`8878dcc` (#52)](https://github.com/morincer/teamcity-plugin-saml/pull/52/commits/8878dccd3ef2513420554cdeb65d19de6d711d59)

Tests still all pass atleast :)